ir/instr.cpp

```diff
@@ -834,6 +834,69 @@ unique_ptr<Instr> ICmp::dup(const string &suffix) const {
   return make_unique<ICmp>(getType(), getName() + suffix, cond, a, b);
 }
 
+FCmp::FCmp(Type &type, string &&name, Cond cond, Value &a, Value &b)
+  : Instr(type, move(name)), a(a), b(b), cond(cond) {}
+
+void FCmp::print(ostream &os) const {
+  const char *condtxt = nullptr;
+  switch (cond) {
+  case FALSE: condtxt = "false "; break;
+  case OEQ:   condtxt = "oeq "; break;
+  case OGT:   condtxt = "ogt "; break;
+  case OGE:   condtxt = "oge "; break;
+  case OLT:   condtxt = "olt "; break;
+  case OLE:   condtxt = "ole "; break;
+  case ONE:   condtxt = "one "; break;
+  case ORD:   condtxt = "ord "; break;
+  case UEQ:   condtxt = "ueq "; break;
+  case UGT:   condtxt = "ugt "; break;
+  case UGE:   condtxt = "uge "; break;
+  case ULT:   condtxt = "ult "; break;
+  case ULE:   condtxt = "ule "; break;
+  case UNE:   condtxt = "une "; break;
+  case UNO:   condtxt = "uno "; break;
+  case TRUE:  condtxt = "true "; break;
+  }
+  os << getName() << " = fcmp " << condtxt << a << ", " << b.getName();
+}
+
+StateValue FCmp::toSMT(State &s) const {
+  auto &[av, ap] = s[a];
+  auto &[bv, bp] = s[b];
+  expr val;
+  switch (cond) {
+  case FALSE: val = false; break;
+  case OEQ:   val = av.foeq(bv); break;
+  case OGT:   val = av.fogt(bv); break;
+  case OGE:   val = av.foge(bv); break;
+  case OLT:   val = av.folt(bv); break;
+  case OLE:   val = av.fole(bv); break;
+  case ONE:   val = av.fone(bv); break;
+  case ORD:   val = av.ford(bv); break;
+  case UEQ:   val = av.fueq(bv); break;
+  case UGT:   val = av.fugt(bv); break;
+  case UGE:   val = av.fuge(bv); break;
+  case ULT:   val = av.fult(bv); break;
+  case ULE:   val = av.fule(bv); break;
+  case UNE:   val = av.fune(bv); break;
+  case UNO:   val = av.funo(bv); break;
+  case TRUE:  val = true; break;
+  }
+  return { expr::mkIf(val, expr::mkUInt(1,1), expr::mkUInt(0,1)), ap && bp };
```

Comment:
you can use `toBVBool()` here instead.

Reply:
btw, do we really need those true & false cases??

Reply:
I think it would be better to turn these to boolean constants during parsing or translating.

Reply:
Agreed. Let's remove those then :)